src/inc/til/rle.h

```diff
@@ -0,0 +1,1066 @@
+            std::pair<T, S> item{ value, length };
+            gsl::span<std::pair<T, S>> span{ &item, 1 };
+            _merge(span.begin(), span.end(), position);
+        }
+
+        // Inserts all values between first and last starting at the given position.
+        template<class Iter>
+        void assign(Iter first, Iter last, const S position = gsl::narrow_cast<S>(0))
+        {
+            //TODO: validate that it doesn't overrun?
+            _merge(first, last, position);
+        }
+
+        constexpr bool operator==(const rle& other) const noexcept
+        {
+            return _size == other._size &&
+                   std::equal(_list.cbegin(), _list.cend(), other._list.cbegin());
+        }
+
+        constexpr bool operator!=(const rle& other) const noexcept
+        {
+            return !(*this == other);
+        }
+
+        /*[[nodiscard]] iterator begin() noexcept
+        {
+            return iterator(_list.begin());
+        }*/
+
+        [[nodiscard]] const_iterator begin() const noexcept
+        {
+            return const_iterator(_list.begin());
+        }
+
+        /*[[nodiscard]] iterator end() noexcept
+        {
+            return iterator(_list.end());
+        }*/
+
+        [[nodiscard]] const_iterator end() const noexcept
+        {
+            return const_iterator(_list.end());
+        }
+
+        /*[[nodiscard]] reverse_iterator rbegin() noexcept
+        {
+            return reverse_iterator(end());
+        }*/
+
+        [[nodiscard]] const_reverse_iterator rbegin() const noexcept
+        {
+            return const_reverse_iterator(end());
+        }
+
+        /*[[nodiscard]] reverse_iterator rend() noexcept
+        {
+            return reverse_iterator(begin());
+        }*/
+
+        [[nodiscard]] const_reverse_iterator rend() const noexcept
+        {
+            return const_reverse_iterator(begin());
+        }
+
+        [[nodiscard]] const_iterator cbegin() const noexcept
+        {
+            return begin();
+        }
+
+        [[nodiscard]] const_iterator cend() const noexcept
+        {
+            return end();
+        }
+
+        [[nodiscard]] const_reverse_iterator crbegin() const noexcept
+        {
+            return rbegin();
+        }
+
+        [[nodiscard]] const_reverse_iterator crend() const noexcept
+        {
+            return rend();
+        }
+
+        std::wstring to_string() const
+        {
+            std::wstringstream wss;
+            wss << std::endl
+                << L"Run of size " << size() << " contains:" << std::endl;
+
+            for (auto& item : _list)
+            {
+                wss << wil::str_printf<std::wstring>(L"[%td for %td]", item.first, item.second) << L" ";
+            }
+
+            wss << std::endl;
+
+            return wss.str();
+        }
+
+    protected:
+        // TODO: get Dustin help to not duplicate this for constness.
+        auto _at(S position, S& applies) const

```

Comment:
Nnnnn.... normally I agree with you but I did this semi-intentionally because it's not visible outside the class, hopefully... the return type from this is horrendously gross and is like a.... `boost::container::vec_iterator<std::pair<_Ty1, _Ty2>*, false>`. 

I have this idea that `boost::container::small_vector` probably isn't the ideal inner-type here either and we'll want to drop in a `std::pmr::vector` as soon as @DHowett works out the kinks in #8787 on how to set up the templates to accept allocators as these are a great place to use a memory pool from my perf traces. 

If this is `auto`, then it'll automatically change correctly to `std::vec_iterator<std::pair<_Ty1, _Ty2>, _Ally>` or whatnot when that one member is flipped out and should propagate through the rest of the template just fine.